src/Compilers/Core/Portable/CommandLine/CommonCompiler.SuppressionDiagnostic.cs

```diff
@@ -0,0 +1,108 @@
+                "SP0001",
+                CodeAnalysisResources.SuppressionDiagnosticDescriptorTitle,
+                CodeAnalysisResources.SuppressionDiagnosticDescriptorMessage,
+                "ProgrammaticSuppression",
+                DiagnosticSeverity.Info,
+                isEnabledByDefault: true);
+
+            private readonly Diagnostic _originalDiagnostic;
+            private readonly string _suppressionId;
+            private readonly LocalizableString _suppressionJustification;
+
+            public SuppressionDiagnostic(
+                Diagnostic originalDiagnostic,
+                string suppressionId,
+                LocalizableString suppressionJustification)
+            {
+                Debug.Assert(originalDiagnostic != null);
+                Debug.Assert(originalDiagnostic.ProgrammaticSuppressionInfo != null);
+                Debug.Assert(!string.IsNullOrEmpty(suppressionId));
+                Debug.Assert(suppressionJustification != null);
+
+                _originalDiagnostic = originalDiagnostic;
+                _suppressionId = suppressionId;
+                _suppressionJustification = suppressionJustification;
+            }
+
+            public override DiagnosticDescriptor Descriptor => s_suppressionDiagnosticDescriptor;
+
+            public override string Id => Descriptor.Id;
+
+            public override string GetMessage(IFormatProvider formatProvider = null)
+            {
+                // Diagnostic '{0}: {1}' was programmatically suppressed by a DiagnosticSuppressor with suppresion ID '{2}' and justification '{3}'
+                var localizableMessageFormat = s_suppressionDiagnosticDescriptor.MessageFormat.ToString(formatProvider);
+                return string.Format(formatProvider,
+                    localizableMessageFormat,
+                    _originalDiagnostic.Id,
+                    _originalDiagnostic.GetMessage(formatProvider),
+                    _suppressionId,
+                    _suppressionJustification.ToString(formatProvider));
+            }
+
+            public override DiagnosticSeverity Severity => DiagnosticSeverity.Info;
+            public override bool IsSuppressed => false;
+            public override int WarningLevel => GetDefaultWarningLevel(DiagnosticSeverity.Info);
+            public override Location Location => _originalDiagnostic.Location;
+            public override IReadOnlyList<Location> AdditionalLocations => _originalDiagnostic.AdditionalLocations;
+            public override ImmutableDictionary<string, string> Properties => ImmutableDictionary<string, string>.Empty;
+
+            public override bool Equals(Diagnostic obj)
+            {
+                var other = obj as SuppressionDiagnostic;
+                if (other == null)
+                {
+                    return false;
+                }
+
+                return Equals(_originalDiagnostic, other._originalDiagnostic) &&
```

Comment:
>return [](start = 16, length = 6)

Consider checking if this is the same instance before doing more expensive comparison.